atl-checker/src/algorithms/global/mod.rs

```diff
@@ -0,0 +1,466 @@
+use crate::edg::{Edge, ExtendedDependencyGraph, HyperEdge, NegationEdge, Vertex};
+use std::cmp::max;
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::fmt::Debug;
+
+// Based on the global algorithm described in "Extended Dependency Graphs and Efficient Distributed Fixed-Point Computation" by A.E. Dalsgaard et al., 2017
+pub struct GlobalAlgorithm<
+    G: ExtendedDependencyGraph<V> + Send + Sync + Clone + Debug + 'static,
+    V: Vertex + Send + Sync + 'static,
+> {
+    edg: G,
+    /// The root vertex, which we discover the graph from
+    v0: V,
+    /// The assignment for all vertices, we use this to keep track  of the current assignment
+    assignment: HashMap<V, bool>,
+    /// The distance of a vertex is represented as the index number of which set the vertex is located in
+    dist: VecDeque<HashSet<V>>,
+}
+
+impl<
+        G: ExtendedDependencyGraph<V> + Send + Sync + Clone + Debug + 'static,
+        V: Vertex + Send + Sync + 'static,
+    > GlobalAlgorithm<G, V>
+{
+    pub fn new(edg: G, v0: V) -> Self {
+        Self {
+            edg,
+            v0,
+            assignment: HashMap::<V, bool>::new(),
+            dist: VecDeque::<HashSet<V>>::new(),
+        }
+    }
+
+    /// Firing the global algorithm, which simply reapplying the F_i function explained in the paper
+    /// until the assignment does not change anymore. Lastly the assignment of the root, v0, is returned
+    pub fn run(&mut self) -> bool {
+        self.initialize();
+
+        let mut components = self.get_components();
+
+        while !components.is_empty() {
+            let component = components.pop_front().unwrap();
+            while self.f(component.clone(), self.assignment.clone()) {}
```

Comment:
You are cloning the component on every iteration !? Use more refereces please 😅

Reply:
You can also avoid cloning the assignments, by having two of them and alternating between them.